static/app/utils/profiling/gl/utils.ts

```diff
@@ -0,0 +1,372 @@
+): ResizeObserver => {
+  const handler: ResizeObserverCallback = entries => {
+    onResize(entries);
+    callback?.();
+  };
+
+  for (const c of canvas) {
+    canvasToDisplaySizeMap.set(c, [c.width, c.height]);
+  }
+
+  const resizeObserver = new ResizeObserver(handler);
+
+  try {
+    // only call us of the number of device pixels changed
+    canvas.forEach(c => {
+      resizeObserver.observe(c, {box: 'device-pixel-content-box'});
+    });
+  } catch (ex) {
+    // device-pixel-content-box is not supported so fallback to this
+    canvas.forEach(c => {
+      resizeObserver.observe(c, {box: 'content-box'});
+    });
+  }
+
+  return resizeObserver;
+};
+
+export function resizeCanvasToDisplaySize(canvas: HTMLCanvasElement): boolean {
+  // Get the size the browser is displaying the canvas in device pixels.
+  const size = canvasToDisplaySizeMap.get(canvas);
+
+  if (!size) {
+    const displayWidth = canvas.clientWidth * window.devicePixelRatio;
+    const displayHeight = canvas.clientHeight * window.devicePixelRatio;
+
+    canvas.width = displayWidth;
+    canvas.height = displayHeight;
+
+    return false;
+  }
+
+  const [displayWidth, displayHeight] = size;
+  // Check if the canvas is not the same size.
+  const needResize = canvas.width !== displayWidth || canvas.height !== displayHeight;
+
+  if (needResize) {
+    // Make the canvas the same size
+    canvas.width = displayWidth;
+    canvas.height = displayHeight;
+  }
+
+  return needResize;
+}
+
+export const Transform = {
+  betweenRect(from: Rect, to: Rect): Rect {
+    return new Rect(to.x, to.y, from.width, from.height).scaledBy(
+      to.width / from.width,
+      to.height / from.height
+    );
+  },
+};
+
+export class Rect {
+  origin: vec2;
+  size: vec2;
+
+  constructor(x: number, y: number, width: number, height: number) {
+    this.origin = vec2.fromValues(x, y);
+    this.size = vec2.fromValues(width, height);
+  }
+
+  static Empty(): Rect {
+    return new Rect(0, 0, 0, 0);
+  }
+
+  static From(rect: Rect): Rect {
+    return new Rect(rect.x, rect.y, rect.width, rect.height);
+  }
+
+  get isEmpty(): boolean {
+    return this.width === 0 && this.height === 0;
+  }
+
+  get x(): number {
+    return this.origin[0];
+  }
+  get y(): number {
+    return this.origin[1];
+  }
+  get width(): number {
+    return this.size[0];
+  }
+  get height(): number {
+    return this.size[1];
+  }
+  get left(): number {
+    return this.x;
+  }
+  get right(): number {
+    return this.left + this.width;
+  }
+  get top(): number {
+    return this.y;
+  }
+  get bottom(): number {
+    return this.top + this.height;
+  }
+
+  toMatrix(): mat3 {
+    // it's easier to display a matrix as a 3x3 array. WebGl matrices are row first and not column first
+    // https://webglfundamentals.org/webgl/lessons/webgl-matrix-vs-math.html
+    // prettier-ignore
+    return mat3.fromValues(
+      this.width, 0, 0,
+      0, this.height, 0,
+      this.x, this.y, 1
+    )
```

Comment:
how do you feel about this?

```suggestion
    const {width: w, height: w, x, y} = this;
    
    // it's easier to display a matrix as a 3x3 array. WebGl matrices are row first and not column first
    // https://webglfundamentals.org/webgl/lessons/webgl-matrix-vs-math.html
    // prettier-ignore
    return mat3.fromValues(
      w, 0, 0,
      0, h, 0,
      x, y, 1
    )
```

Reply:
Yup, looks much more readable :) (I'm not a fan of the extra allocations, but who cares)